### ISSUE TITLE
client->server request: Join Lines

### ISSUE THREAD

**Issue by matklad** [microsoft/language-server-protocol#992]
Meta: I am in the process of formalizing protocol extensions, used by rust-analyzer. I will be filing issues for generally useful extensions. Here's [the doc](https://github.com/rust-analyzer/rust-analyzer/blob/5b5ebec440841ee98a0aa70b71a135d94f5ca077/docs/dev/lsp-extensions.md) where we will document our extensions. 

## `joinLines`

**Server Capability:** `{ "joinLines": boolean }`

This request is send from client to server to handle "Join Lines" editor action.

**Method:** `experimental/JoinLines`

**Request:**

```typescript
interface JoinLinesParams {
    textDocument: TextDocumentIdentifier,
    /// Currently active selections/cursor offsets.
    /// This is an array to support multiple cursors.
    ranges: Range[],
}
```

**Response:**

```typescript
TextEdit[]
```

### Example

```rust
fn main() {
    /*cursor here*/let x = {
        92
    };
}
```

`experimental/joinLines` yields (curly braces are automagiacally removed)

```rust
fn main() {
    let x = 92;
}
```

### Unresolved Question

* What is the position of the cursor after `joinLines`?
  Currently this is left to editor's discretion, but it might be useful to specify on the server via snippets.
  However, it then becomes unclear how it works with multi cursor.

**Comment by dbaeumer** [comment]
@jrieken FYI.

**Comment by rwols** [comment]
I'm not sure if I would like a text-oriented operation to be dependent on a language server. Why can't I use the regular join lines followed by a code action?

**Comment by matklad** [comment]
Using code action is certainly possible (you don't even need regular join lines, you can fold everything into code action). However in my experience as a user of IntelliJ, dwim "text-oriented" commands (enter, open line, join lines), if implemented correctly, can create a significantly better flow of development. Here's a nice example from Enter in Kotlin:

![hello](https://user-images.githubusercontent.com/1711539/85343267-b9c1f880-b4ec-11ea-8507-d427c608e49b.gif)


I am pretty sure that there will be users who'd love this, and users who'd hate this. My gut feeling is that the first group is bigger. 

From the protocol point of view, I'd think that we should unlock opportunities to precisely tweak user experience, rather than prescribe  a specific way of doing things. Severs are free to not use the opportunities, users are free to disable specific features they feel are getting in the way.

**Comment by rwols** [comment]
I get that it's cool to have that functionality, but your above GIF would be implementable by a simple snippet.

**Comment by matklad** [comment]
> but your above GIF would be implementable by a simple snippet.

I strongly disagree here: snippets indeed can be used to insert the same text, but it’s not the point. The point is dwim experience during basic editing operations, and that, by definition, can not be replaced with explicit user actions. 

See also https://github.com/JetBrains/kotlin/tree/master/idea/src/org/jetbrains/kotlin/idea/joinLines which defined Kotlin’s custom behavior for join lines. There’s quite a bit of code there, and it can’t be replaced by simple snippets/code actions/etc.

It can be argued that having those features is not important/detrimental, but that would have to be a different argument from “we already can simulate then we’ll enough”.

**Comment by puremourning** [comment]
I don't know if it matters to you but i'm almost 100% certain i would not implement this in my client.
